### ISSUE TITLE
Add type definition file for hydra functions

### ISSUE THREAD

**Issue by ojack** [ojack/hydra-synth#100]
Would be useful for code completion in editors, as discussed at https://github.com/ojack/hydra/issues/109

I am not familiar with TypeScript or *.ts files, so especially appreciate inputs / developments / pull requests related to this.

Related to  https://github.com/folz/hydra-ts .

**Comment by bodobraegger** [comment]
Hey, I am currently developing an extension for the Obsidian editor, to write and render hydra snippets inline. For this, I wrote my own type definitions, and I can create PR soon to merge them back into this repo, forking it as we speak.

thanks!!

**Comment by bodobraegger** [comment]
I could also include a very basic typescript testing setup. It is important to note that none of the proposed changes would affect the compiled project, it is only to make development and integrating with TS projects easier.

Including the testing setup would add typescript to the devDependencies.

**Comment by bodobraegger** [comment]
I also ran `npm update` in the process. Should I seperate this out into a prior pull request?

**Comment by ojack** [comment]
Oh this sounds great! Let's do the npm update in a separate PR :)

**Comment by bodobraegger** [comment]
Sweet! That one is already up and ready to merge as far as I can tell, #168.

I need to fix a bug in the type declarations for relative imports, will reopen a PR from that branch when it's ready. 

Thanks!:)

**Comment by bodobraegger** [comment]
Hey, as far as I can tell, the type declaration does not work with the CJS / require syntax. I wrote a patch to use the ECM version again, but I am not sure if this breaks other projects. What do you think @ojack ?

**Comment by therealmainmethod** [comment]
@bodobraegger @ojack Are there any new Updates on this Issue? I see @bodobraegger already had type declarations included in his pull request but this seems to be not included yet. 

I'm also looking out for this enhancement, as I'm using hydra in a lot of typescript related projects.

**Comment by ojack** [comment]
Hi sorry, I got lost on the current status of this. Let me know if there is something ready to review.

@therealmainmethod you can also try https://github.com/folz/hydra-ts !

**Comment by jdomizz** [comment]
> **Note:** This comment has been revised to match the actual PR #211, which contains 3 `.d.ts` files instead of the 12 originally proposed.

Following up on the discussion here and @bodobraegger's proposal — here's what's in the PR and the reasoning behind the approach.

---

**TL;DR:**
- 3 `.d.ts` files alongside the source (zero changes to runtime code).
- Creative coders get autocompletion by referencing `global.d.ts`.
- No new dependencies, no build step.

---

## What's in the branch

Branch `feat/types` on `jdomizz/hydra-synth` adds **3 `.d.ts` files**. Zero changes to the JS. No new dependencies. No build step.

| File | What it declares |
|------|-----------------|
| `src/global.d.ts` | The live-coding DSL: generators (`osc`, `noise`, `shape`, ...), render targets (`Output` interface for `o0`–`o3`), sources (`HydraSource` interface for `s0`–`s3`), audio (`Audio`, `AudioOpts`), `Param`, `SourceInput`, and the rhythmic `Array` extensions (`fast`, `smooth`, `ease`, ...). Self-contained — no imports, no triple-slash references — so it works in a folder without `node_modules`. |
| `src/hydra-synth.d.ts` | The library API: the `HydraRenderer` class (exported as default), `HydraOpts` (constructor options), and the `Synth` context. References `global.d.ts` for the DSL types. |
| `src/glsl/glsl-functions.d.ts` | The transforms catalog: `GlslFunction`, `GlslInput`, `GlslFunctionType`. |

`package.json` adds a `"types"` field and `exports.types` conditions so the declarations resolve automatically for npm consumers.

## Why `.d.ts` and not JSDoc

JSDoc works, but at hydra-synth's scale it has three concrete problems:

1. **Noise ratio.** Every public property needs a `@type` annotation — hundreds of lines of comments competing visually with the logic.
2. **Branch duplication.** Same property, same type, declared twice because it's in an `if/else`.
3. **Maintenance friction.** When you change a function signature, the JSDoc is in the same file as the implementation. Makes diffs harder to scan.

With `.d.ts`, the JS stays clean. The types are in separate files — one declaration per symbol, no branches, no repetition. When you change an API, you edit one line in a `.d.ts` file that has nothing but signatures.

## Two entry points

| Audience | Mechanism | Key file |
|----------|-----------|----------|
| **npm consumer** (`npm install hydra-synth`, `import Hydra from 'hydra-synth'`) | Automatic via `"types"` in `package.json`. | `src/hydra-synth.d.ts` |
| **Creative coder** (CDN `<script>`, no npm) | Download `global.d.ts` from unpkg/GitHub, drop it next to your sketch. | `src/global.d.ts` |

`global.d.ts` is deliberately self-contained (no imports) so CDN users can reference it with a single `/// <reference path>` — no generated file, no build step to keep in sync. Editor plugins can also inject it via `jsconfig.json` / `tsconfig.json`.

The third file, `src/glsl/glsl-functions.d.ts`, covers the `glsl-functions` subpath export (public since 1.3.28, #141) — typed metadata about the transform catalog for plugin/extension authors who need to introspect available transforms programmatically.

## The `Array.prototype` trade-off

`Array.prototype` extensions (`.fast()`, `.smooth()`, `.ease()`, `.offset()`, `.fit()`) are declared as global interface augmentations. The consequence: **every** array in the project sees these methods in autocompletion, not just hydra arrays. This is inherent to mutating `Array.prototype` — there's no way to scope type augmentations to "only in hydra code." The alternatives would be:

- Don't declare them → every sketch using arrays gets type errors.
- Move to a wrapper (e.g. `seq([1,2,3]).fast(2)`) → cleaner types, but changes runtime behavior and breaks existing sketches.

I went with the global declaration and documented the trade-off.

## What's NOT typed

Deliberately excluded from the declarations:

| File | Reason |
|------|--------|
| `glsl-functions.js` | ~50 generated transform definitions — already typed transitively via `glsl-functions.d.ts` |
| `glsl/renderpass-functions.js` | Single renderpass (`blur`) using `glslify`, not part of the live-coding DSL |
| `glsl/utility-functions.js` | Internal GLSL helpers (`_luminance`, `_noise`, …) used by `glsl-source.js` at shader compile time — not exposed to consumers |
| `generate-glsl.js` | Internal GLSL codegen, no consumer value |
| `shader-generator.js` | Standalone GLSL code generator — captures shader output as a string instead of rendering to a canvas; not wired into the runtime |
| `shaderManager.js` | Dead code — not imported anywhere; appears to be an early fragment-shader prototype |
| `index.js` | CJS entrypoint (`require('hydra-synth')`) — re-exports `HydraRenderer`, no own API |

## Notes on `Param` and `sum()`

`Param` is the type used for dynamic shader arguments — any scalar value that a sketch passes to a transform (e.g. `brightness(0.5)`, `rotate(() => time * 0.1)`, `[0, 1].fast(2)`). It resolves to `number | (() => number) | number[]` at runtime.

`Param` does not accept a `GlslSource`: the "collapse a chain to a scalar via `sum`" path in `format-arguments.js` never produced valid GLSL (`sum` returns `vec4`, which would land in a `float` slot), so the types describe what actually works. Related to #204.

Plain arrays passed to vector inputs throw `Arguments must be a texture or GlslSource` (e.g. the documented `.sum([1,1,1,1])`). `sum`'s `scale` is therefore typed as `SourceInput` — what the runtime accepts today. Also related to #204.

## Verification

The declarations were verified against the runtime with `tsc --strict` in both usage modes (npm import + CDN reference), plus shader generation for every function in the catalog.

## What this means for maintenance

When you change a public API — add a method, change a parameter, remove something — edit **one line** in the corresponding `.d.ts` file:

```ts
// Method: name(params): returnType
osc(frequency?: number, sync?: number, offset?: number): GlslSource;

// Property on a class:
canvas: HTMLCanvasElement;
```

That's the whole workflow — no type annotations in the JS, no build step, no new dependencies.

**Comment by bodobraegger** [comment]
I think this looks good to me! Maybe we can add pre-commit hook to scan for API changes and see if the types file is up to date? Otherwise, I fear that we would accumulate drift quickly by developers not used to it. Also perhaps mention the typing system in the readme.

**Comment by ojack** [comment]
Yes this looks great! @jdomizz please do open a PR with this. We can add a short note somewhere in `README.md`.

I like the pre-commit hook idea, but would consider it outside the scope of this specific PR.

**Comment by jdomizz** [comment]
@bodobraegger I agree that the pre-commit hook is a good idea. I’ve left it out of this pull request, as @ojack suggested.

**Comment by bodobraegger** [comment]
Great! Thanks a ton. I can take a look at the pre-commit if you haven’t already done it.

**Comment by jdomizz** [comment]
@bodobraegger @ojack Before you dive into pre-commit hooks — here's a thought on how to introduce this incrementally without overwhelming the repo with tooling it doesn't have yet.

**Step 1 — Just npm scripts**
Add `typecheck` (`tsc --noEmit`) to `package.json` along with a `tsconfig.json` and 3 test files: 
- `types-test/library.ts` — library API (`import Hydra from 'hydra-synth'`)
- `types-test/sketch.ts` — live-coding DSL (globals like `osc`, `s0`, `speed`)
- `types-test/restrictions.ts` —  `@ts-expect-error` assertions for intentional restrictions 

The type tests are small `.ts` files that consume the `.d.ts` declarations exactly how a real user would, `tsc` typechecks them via the `include` in `tsconfig.json`. Anyone who wants can run it locally, but nothing is enforced. The maintainer can run it before merging.

**Step 2 — Pre-commit hook (optional)**
If the team wants a convenience before pushing, husky + a `pre-commit` that runs `typecheck` is lightweight and well-tested. But it can be skipped with `--no-verify` and doesn't run in CI — not a real safety net, just a local convenience.

**Step 3 — CI workflow (when the team is ready)**
A single GitHub Actions job (`npm ci && npm run typecheck`) is ~20 lines of YAML. This is the only option that can't be skipped, works for every PR, and doesn't depend on anyone remembering to run anything. It's also the foundation for future checks (build diff, vitest, linting).

**Bonus — catalog coverage script (`typecheck:catalog`)**
As an optional extra, a Node script could compare the runtime transform catalog in `glsl-functions.js` against the `.d.ts` declarations by AST, so that adding a transform without typing it breaks the build. Pros: catches drift that `tsc` alone won't spot. Cons: real AST parsing code to maintain, and for a repo with one maintainer and sporadic changes it may be overkill. Could be added later if drift becomes a pattern.

**The tradeoff:** each step adds real value, but each step also adds a thing the team has to maintain and understand. Better to land the typings, prove they're useful in daily workflow, then harden. 

Happy to discuss whichever approach fits the team best.
